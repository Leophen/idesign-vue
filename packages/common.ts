import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
