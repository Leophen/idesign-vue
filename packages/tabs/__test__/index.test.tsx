import { vi } from 'vitest';
import { Tabs, TabsItem } from '../index';
import { render } from '../../common'

describe('Tabs 组件测试', () => {
  it('create', () => {
    const wrapper = render(
      <Tabs>
        <TabsItem>选项卡1</TabsItem>
        <TabsItem>选项卡2</TabsItem>
        <TabsItem>选项卡3</TabsItem>
      </Tabs>
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('active', () => {
    const wrapper = render(
      <Tabs modelValue={1}>
        <TabsItem>选项卡1</TabsItem>
        <TabsItem>选项卡2</TabsItem>
        <TabsItem>选项卡3</TabsItem>
      </Tabs>
    );
    expect(wrapper.findAll('.i-tabs-item')[1].classes()).toContain('i-tabs-item__active');
  });

  it('defaultActive', () => {
    const wrapper = render(
      <Tabs defaultActive={1}>
        <TabsItem>选项卡1</TabsItem>
        <TabsItem>选项卡2</TabsItem>
        <TabsItem>选项卡3</TabsItem>
      </Tabs>
    );
    expect(wrapper.findAll('.i-tabs-item')[1].classes()).toContain('i-tabs-item__active');
  });

  it('onChange', () => {
    const changeFn = vi.fn();
    const wrapper = render(
      <Tabs onChange={changeFn}>
        <TabsItem>选项卡1</TabsItem>
        <TabsItem>选项卡2</TabsItem>
        <TabsItem>选项卡3</TabsItem>
      </Tabs>
    );
    wrapper.findAll('.i-tabs-item')[1].trigger('click');
    expect(changeFn).toHaveBeenCalled();
  });

  it('theme', () => {
    const wrapper = render(
      <Tabs theme="card">
        <TabsItem>选项卡1</TabsItem>
        <TabsItem>选项卡2</TabsItem>
        <TabsItem>选项卡3</TabsItem>
      </Tabs>
    );
    expect(wrapper.find('.i-tabs-item__card').exists()).toBe(true);
  });

  it('disabled', () => {
    const changeFn = vi.fn();
    const wrapper = render(
      <Tabs disabled>
        <TabsItem>选项卡1</TabsItem>
        <TabsItem data-testid='test-item'>选项卡2</TabsItem>
        <TabsItem>选项卡3</TabsItem>
      </Tabs>
    );
    wrapper.findAll('.i-tabs-item')[1].trigger('click');
    expect(changeFn).not.toHaveBeenCalled();
  });
});

