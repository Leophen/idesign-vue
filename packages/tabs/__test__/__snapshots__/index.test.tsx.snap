// Vitest Snapshot v1

exports[`Tabs 组件测试 > create 1`] = `
<div
  class="i-tabs"
>
  
  <div
    class="i-tabs-item i-tabs-item__active"
    data-active="true"
    data-disabled="false"
  >
    
    
    选项卡1
    
    
  </div>
  <div
    class="i-tabs-item"
    data-active="false"
    data-disabled="false"
  >
    
    
    选项卡2
    
    
  </div>
  <div
    class="i-tabs-item"
    data-active="false"
    data-disabled="false"
  >
    
    
    选项卡3
    
    
  </div>
  
  <div
    class="i-tabs__bar"
    style="width: 0px; left: 0px;"
  />
</div>
`;
